### ISSUE TITLE
Make initialValues optional, touch Field on submit

### ISSUE THREAD

**Issue by charlax** [jaredpalmer/formik#691]
**Is your feature request related to a problem? Please describe.**

For extra context, see #445

I'm trying to port some of my simple forms from `redux-form` to `formik`. One first issue I had was that `initialValues` is required.

I'm using some HOC that creates full-fledged forms from a component describe only the form fields (e.g. for a user profile: first name, last name, password, etc.). I find that it is duplicate work to define all fields in the form component, and then to redefine all those keys with an empty value (most often an empty string `''`) - especially since `redux-form` gets away without it.

Something that looks like this:

```javascript
// UserProfilePage.js

import UserProfileForm from './UserProfileForm'

// ...

export default createPage({component: UserProfileForm, url: '/users/self', ... })
```

```javascript
// UserProfileForm.js

export default const UserProfileFrom = () => (
  <div>
    <Field name="firstname" label="First name" /> 
     ...
  </div>
);

```

Formik would have me do this instead:

```javascript
// UserProfilePage.js

// ...

export default createPage({
  component: UserProfileForm,
  url: '/users/self',
  initialValues: { firstname: '', lastname: '' },
  // ...
});
```

... which is pretty redundant.

**Describe the solution you'd like**

I think `initialValues` and `mapPropsToValues` could be made optional, defaulting all values to an empty string `''`.

On top of that, fields defined using `Field` could be registered so that when the form is submitted, they are marked as touched. Currently, they are not, which means that any error that is for a field that wasn't explicitly defined in `initialValues` or `mapPropsToValues` will not be shown - you can see an example here: https://codesandbox.io/s/1o7lkpmk8j

**Describe alternatives you've considered**

For the second issue (auto-touching `Field`), an alternative could be that the developer sets all the keys returned in the errors as touched, but that seems very hacky.

Let me know what you think!

**Comment by jaredpalmer** [comment]
This will now be possible thanks to #684 . We'll now be able to construct initialValues (if not specified) by looking at what fields have mounted and assuming they are just strings.

**Comment by Andreyco** [comment]
@charlax hey, this is now possible with beta 5! We should probably mark `initialValues` optional for Typescript users (you still have to pass at least empty object) but I am afraid people start to complain why fields are not marked as touched (when not using `Field` component)

**Comment by peteruithoven** [comment]
Is there an update on this? I'm running version `1.0.1`, but when I don't provide the `initialValues` I still get the changed an input from uncontrolled to controlled error. (I'm using the Render prop)

**Comment by wachunga** [comment]
It would be really nice to have a default initialValues. 👍

**Comment by charlax** [comment]
Still think this would be important :)

**Comment by charlax** [comment]
The [doc says](https://jaredpalmer.com/formik/docs/guides/form-submission) that formik touches all fields, but I am not observing this for `FieldArray`. Is that expected?

**Comment by charlax** [comment]
Any help needed here? Would be happy to have a look at this since I need it. With 1.4.0, it's still not auto-touching registered fields.

**Comment by charlax** [comment]
For the moment, I'm just using `submitCount`:

```jsx
const touched = getIn(form.touched, field.name) || form.submitCount > 0;
```

**Comment by charlax** [comment]
Another issue: when submitting then resetting the form, it seems that the `initialValues` are not reset as well.

**Comment by nicoponder** [comment]
Is this going to be addressed? Recent activity here and in other issues/PRs doesn't say one way or the other, or is stale.

Using `submitCount` to delay showing validation errors whenever `initialValues` can't be fully provided works just fine so I don't see a pressing need to, but it does feel like a bug when you bump into this behavior. It'd definitely be much nicer if we could rely on all fields to be touched on submit.

**Comment by lukewlms** [comment]
My team is new to Formik - does this issue mean that every time we initialize a form we have to explicitly set every field in initialValues to empty?

That seems like a lot of unnecessary typing, but also does seem like the way the library is supposed to be used, if not using one the workarounds [here](https://github.com/jaredpalmer/formik/issues/691#issuecomment-446509600) or [here](https://github.com/jaredpalmer/formik/issues/738#issuecomment-408056088).

**Comment by isaacalves** [comment]
I could go deeper investigating this but I’ve spent way too much time already, also I'm a bit exhausted. Here's the outcome of my research and how I fixed these problems:

If a form is not dirty (no fields are touched) and the user tries to submit, the ErrorMessage might not be shown. This might be because field is not touched. So, even though it has errors, the error messages won’t be displayed because the fields aren't touched. That makes sense to me unless _no fields_ are touched, in which case all fields with errors should be “touched” by Formik so the ErrorMessages are displayed, or something like that.

It might be that the fields aren’t set to touched because they don’t have an initialValue, so always set an initial value, an empty string works. Not sure why but that’s how it works and it was suggested by the author on one of these threads.

In one use case of mine, a form is rendered conditionally on a page with another form (like Step 1 and 2) - _I think_ that for this reason the initialValues didn’t work, but I might be wrong. So, even though initialValues were set, when debugging the props, initialValues would just be an empty object. And therefore, trying to submit the form without the user touching the fields wouldn't display the error messages (because Formik wouldn't set them to touched). But using `enableReinitialize={true}` in the Formik component fixed that (also not sure why)

On my project I have a bunch of forms and I'm not using the Field component, just a custom Input, so maybe if I use Field these problems won't happen - I might do that at some point. I'm also using Yup for validation.

**Comment by h0jeZvgoxFepBQ2C** [comment]
I have the same usecase as @isaacalves - imo the error messages should be shown anyway, touched or not touched. I really don't see the point in it, to show it only when it's touched... The need to set empty defaultValues and touch every field is really cumbersome :(((((

**Comment by Vechtomov** [comment]
Maybe it will help someone. I use this instead of useField.
```
export function useSelfRegisteredField(propsOrFieldName) {
  const formContext = useFormikContext();
  const [field, meta] = useField(propsOrFieldName);
  useEffect(() => {
    if (field.value === undefined)
      formContext.setFieldValue(field.name, '', false);
  }, []);
  return [field, meta];
}
```

**Comment by Dhruv26** [comment]
This seems to be an issue in the case of `<FieldArray />` because you cannot specify the initial value for Fields that you dynamically add to the array. Is there a workaround for this?

**Comment by jaredpalmer** [comment]
I agree with touching all registered fields on submit and defaulting to empty string.

**Comment by helgefredheim** [comment]
@jaredpalmer I add fields to my form when the user fills it out, depending on what s/he selects. The fields which are added are not marked as touched when the user submits the form; thus no error messages are displayed. Is there a way to get around this?

**Comment by jaredpalmer** [comment]
Correct this is why this change is necessary

**Comment by thilinaucsc** [comment]
also not submit that value to thought backend.

**Comment by jasonhendricksdev** [comment]
Any update for this?

Thanks in advance

**Comment by DannerTerra** [comment]
Any progress on this? or #684 ?

**Comment by odama626** [comment]
Just ran into this and since I've been using useField I did something a little different

```
import { useField as formikUseField, useFormikContext } from 'formik'

function useField(props) {
  const fieldProps = formikUseField(props);
  const form = useFormikContext();
  fieldProps[1].touched = fieldProps[1].touched || form.submitCount > 0;
  return fieldProps;
}
```

**Comment by HiranmayaGundu** [comment]
@jaredpalmer Is anyone working on this? If not I don't mind giving it a shot, if you could point me in the right direction!

**Comment by photonite** [comment]
The problem with touching an item in a FieldArray based on submitCount is that a newly added item just after one submit will be automatically touched, and thus the validation errors will be visible which I think is not the desired behavior.
